### ISSUE TITLE
Registry v2 as pull-through cache eating up CPU

### ISSUE THREAD

**Issue by schmunk42** [docker/distribution#944]
I configured a v2 registry mirror or pull-through cache like so:

```
registry2:
  image: registry:2
  ports:
    - 5000:5000
  links:
    - redis:REDIS
  environment:
    - REGISTRY_PROXY_REMOTEURL="https://registry-1.docker.io"
    - REGISTRY_REDIS_ADDR="REDIS:6379"
redis:
  image: redis
  expose:
    - 6379
  ports:
    - 6379:6379
```

I tried it also with host-volumes and without redis and `inmemory` settings, various versions (2, 2.1, latest); but all results are the same.

A few sequential pulls might work, but at least on two concurrent pulls, which have to be fetched by the registry, the registry container runs into some kind of idle state, eating up 70-100% of the CPU - eventually dropping out a few kB/sec of data.

Here's a longer gist of the log output https://gist.github.com/schmunk42/aeede32b28d3e94e4d30.

*Using: docker 1.8.1 with docker-machine VMs (Toolbox)*

**Comment by heww** [comment]
Same problem for me, and I will got a error `Unsolicited response received on idle HTTP channel starting with "H"; err=<nil>`

**Comment by RichardScothern** [comment]
Thanks for the reports.  Fix is on its way..

**Comment by roboll** [comment]
As far as I can tell this makes registry completely useless as a cache - is there any progress on getting a fix released?

**Comment by RichardScothern** [comment]
There is a fix referenced above (PR #1006).  This will make it into the next registry version.
